### DISCUSSION TITLE
Bug: Switching to IL2CPP is giving a JsonSerializationException

### DISCUSSION THREAD

**Discussion by jkafkaris** [applejag/Newtonsoft.Json-for-Unity#69] [Q&A] [answered]
I am using a .net library (https://github.com/contentful/contentful.net) that loads json from a CMS online. I am also using this json library. Using Mono it works perfect. 

As one of the project's targets is Android, Google has requested after it was submitted that it must be compliant with the Play 64-bit requirement.

It goes on to instruct to change the scripting backend to IL2CPP and check ARM64. When building and testing on Android the following error I think occurs when it hits this logic:

```cs
json.SelectToken("$.items").ToObject<IEnumerable<T>>(Serializer);
```

error that occurs:

> ``Newtonsoft.Json.JsonSerializationException: Error converting value "Hero" to type 'System.Collections.Generic.Dictionary`2[System.String,System.String]'. Path 'fields.title'. ---> System.ArgumentException: Could not cast or convert from System.String to System.Collections.Generic.Dictionary`2[System.String,System.String].``

This is the point in the json that it appears to be falling at - which is right near the start of the data:

```json
      "fields": {
        "title": "Hero",
        "file": {
          "url": "//images.hero.jpg",
          "details": {
            "size": 10000,
            "image": {
              "width": 1024,
              "height": 768
            }
          },
          "fileName": "Hero.jpg",
          "contentType": "image/jpeg"
        }
      }
```

## Expected behavior

For the app to run as it does in the editor or Mono

## Actual behavior

Android build with IL2CPP and ARM64 gives the above error.

## Steps to reproduce

Unsure how to simplify a reproduction of the issue.

## Details

Developed on Windows 10
Targeting Android API Level 30
Android version 2019.3.2f1
Newtonsoft.Json-for-Unity package version 12.0.301

## Checklist

- [x] Shutdown Unity, deleted the /Library folder, opened project again in Unity, and problem still remains.
- [X] Checked to be using latest version of the package.

Thanks.

**Comment by applejag** [comment]
Hi @jkafkaris ! Thanks for reaching out about this!

I've inspected this a bit but I can't come with any answers. I'm missing too much context.

Would it be possible if you could send the entire JSON text you're serializing (or as little as possible but still triggering the issue) as well as the type models? What is `T` in `IEnumerable<T>`? If you could give the whole tree of classes/structs that you're deserializing, the would be great!

**Comment by jkafkaris** [comment]
Hi @jilleJr 

Thanks for getting back to me! I'll be back on the project on Monday and will try collate some of the data you have requested and get back to you.

**Comment by jkafkaris** [comment, marked as answer]
Hi @jilleJr 

Good news, I needed to include a link.xml file to stop Unity stripping out relevant code from the Contentful assembly.

I had it more narrowed but in the end preserved the whole assembly.

`<linker>
    <assembly fullname="ContentfulAssembly" preserve="all" />
</linker>`

Works as expected now.
